numpy/ma/core.pyi

```diff
@@ -447,15 +461,68 @@ masked_print_option: _MaskedPrintOption
 
 def flatten_structured_array(a): ...
 
-class MaskedIterator:
-    ma: Any
+class MaskedIterator(Generic[_ShapeT_co, _DTypeT_co]):
+    ma: MaskedArray[_ShapeT_co, _DTypeT_co]
     dataiter: Any
     maskiter: Any
-    def __init__(self, ma): ...
-    def __iter__(self): ...
-    def __getitem__(self, indx): ...
-    def __setitem__(self, index, value): ...
-    def __next__(self): ...
+    def __init__(self, ma: MaskedArray[_ShapeT_co, _DTypeT_co]) -> None: ...
+    def __iter__(self) -> Iterator[Any]: ...
+
+    # Similar to `MaskedArray.__getitem__` but without the `void` case.
+    @overload
+    def __getitem__(self, indx: _ArrayInt_co | tuple[_ArrayInt_co, ...], /) -> MaskedArray[_AnyShape, _DTypeT_co]: ...
+    @overload
+    def __getitem__(self, indx: SupportsIndex | tuple[SupportsIndex, ...], /) -> Any: ...
+    @overload
+    def __getitem__(self, indx: _ToIndices, /) -> MaskedArray[_AnyShape, _DTypeT_co]: ...
+
+    # Similar to `ndarray.__setitem__` but without the `void` case.
+    @overload  # flexible | object_ | bool
+    def __setitem__(
+        self: MaskedIterator[Any, dtype[flexible | object_ | np.bool] | dtypes.StringDType],
+        index: _ToIndices,
+        value: object,
+        /,
+    ) -> None: ...
+    @overload  # integer
+    def __setitem__(
+        self: MaskedIterator[Any, dtype[integer]],
+        index: _ToIndices,
+        value: _ConvertibleToInt | _NestedSequence[_ConvertibleToInt] | _ArrayLikeInt_co,
+        /,
+    ) -> None: ...
+    @overload  # floating
+    def __setitem__(
+        self: MaskedIterator[Any, dtype[floating]],
+        index: _ToIndices,
+        value: _ConvertibleToFloat | _NestedSequence[_ConvertibleToFloat | None] | _ArrayLikeFloat_co | None,
+        /,
+    ) -> None: ...
+    @overload  # complexfloating
+    def __setitem__(
+        self: MaskedIterator[Any, dtype[complexfloating]],
+        index: _ToIndices,
+        value: _ConvertibleToComplex | _NestedSequence[_ConvertibleToComplex | None] | _ArrayLikeNumber_co | None,
+        /,
+    ) -> None: ...
+    @overload  # timedelta64
+    def __setitem__(
+        self: MaskedIterator[Any, dtype[timedelta64]],
+        index: _ToIndices,
+        value: _ConvertibleToTD64 | _NestedSequence[_ConvertibleToTD64],
+        /,
+    ) -> None: ...
+    @overload  # datetime64
+    def __setitem__(
+        self: MaskedIterator[Any, dtype[datetime64]],
+        index: _ToIndices,
+        value: _ConvertibleToDT64 | _NestedSequence[_ConvertibleToDT64],
+        /,
+    ) -> None: ...
+    @overload  # catch-all
+    def __setitem__(self, index: _ToIndices, value: ArrayLike, /) -> None: ...
+
+    def __next__(self) -> Any: ...
```

Comment:
similar to

https://github.com/numpy/numpy/blob/f5a6af86acab7fcd1644fad76b5fbe466a0a98dd/numpy/ma/core.pyi#L576-L577

https://github.com/numpy/numpy/blob/00bbfefed97140b3c493b9154d4db62dfe2cf787/numpy/__init__.pyi#L2112-L2113

If that's `Any`, I don't think we can narrow this one down more? And if we can, we should probably update that one as well / first?

Reply:
Perhaps it'd help if we change this to `-> Incomplete`